tiledb/sm/storage_manager/consolidator.cc

```diff
@@ -190,20 +190,90 @@ Status Consolidator::consolidate_fragments(
     uint32_t key_length) {
   STATS_START_TIMER(stats::GlobalStats::TimerType::CONSOLIDATE_FRAGS)
 
-  // Get array schema
-  URI array_uri = URI(array_name);
-  EncryptionKey enc_key;
-  RETURN_NOT_OK(enc_key.set_key(encryption_type, encryption_key, key_length));
-  auto array_schema = (ArraySchema*)nullptr;
-  RETURN_NOT_OK(
-      storage_manager_->load_array_schema(array_uri, enc_key, &array_schema));
+  // Open array for reading
+  Array array_for_reads(URI(array_name), storage_manager_);
+  RETURN_NOT_OK(array_for_reads.open_without_fragments(
+      encryption_type, encryption_key, key_length));
 
-  // Consolidate fragments
-  RETURN_NOT_OK_ELSE(
-      consolidate(array_schema, encryption_type, encryption_key, key_length),
-      tdb_delete(array_schema));
+  // Open array for writing
+  Array array_for_writes(array_for_reads.array_uri(), storage_manager_);
+  RETURN_NOT_OK(array_for_writes.open(
+      QueryType::WRITE, encryption_type, encryption_key, key_length));
+
+  // Get fragment info
+  FragmentInfo fragment_info;
+  auto st = storage_manager_->get_fragment_info(
+      array_for_reads,
+      config_.timestamp_start_,
+      config_.timestamp_end_,
+      &fragment_info);
+  if (!st.ok()) {
+    array_for_reads.close();
+    array_for_writes.close();
+    return st;
+  }
+
+  uint32_t step = 0;
+  FragmentInfo to_consolidate;
+  do {
+    // No need to consolidate if no more than 1 fragment exist
+    if (fragment_info.fragment_num() <= 1)
+      break;
+
+    // Find the next fragments to be consolidated
+    NDRange union_non_empty_domains;
+    st = compute_next_to_consolidate(
+        array_for_reads.array_schema(),
+        fragment_info,
+        &to_consolidate,
+        &union_non_empty_domains);
+    if (!st.ok()) {
+      array_for_reads.close();
+      array_for_writes.close();
+      return st;
+    }
+
+    // Check if there is anything to consolidate
+    if (to_consolidate.fragment_num() <= 1)
+      break;
+
+    // Consolidate the selected fragments
+    URI new_fragment_uri;
+    st = consolidate(
+        array_for_reads,
+        array_for_writes,
+        to_consolidate,
+        union_non_empty_domains,
+        &new_fragment_uri);
+    if (!st.ok()) {
+      array_for_reads.close();
+      array_for_writes.close();
+      return st;
+    }
+
+    // Get fragment info of the consolidated fragment
+    SingleFragmentInfo new_fragment_info;
+    st = storage_manager_->get_fragment_info(
```

Comment:
This still has a problem in that `get_fragment_info` calls `array_reopen` which ends up doing a full listing again and reloading the metadata when most of it has not changed. The `new_fragment_uri` is only filtered after doing a full listing in this code path.

Reply:
Disregard please.